web/src/settings_components.ts

```diff
@@ -395,20 +396,30 @@ function read_select_field_data_from_form(
     $profile_field_form.find("div.choice-row").each(function (this: HTMLElement) {
         const text = $(this).find("input")[0].value;
         if (text) {
-            if (old_option_value_map.get(text) !== undefined) {
-                // Resetting the data-value in the form is
-                // important if the user removed an option string
-                // and then added it back again before saving
-                // changes.
-                $(this).attr("data-value", old_option_value_map.get(text));
-            }
             const value = $(this).attr("data-value")!;
-            field_data[value] = {text, order: field_order.toString()};
+            field_data.set(value, {text, order: field_order.toString()});
             field_order += 1;
         }
     });
 
-    return field_data;
+    // Check if the user has deleted an option and added the same option again by manually
+    // filling out its value. If that's the case, we replace the new key in `field_data`
+    // with the original key from `old_option_value_map`. This key is what is used on the
+    // user side to specify their choice in a list, so we don't want to change the key if
+    // the text has not changed.
+    for (const [old_choice, old_key] of old_option_value_map.entries()) {
+        // We want to find the first occurrence of the old choice, if there are multiple
+        // occurrences, we want to let the other occurrences be as is, since we rely on
+        // the API response to give us an error in case of duplicate o_ption values.
+        const deleted_choice = [...field_data].find(([, value]) => value.text === old_choice);
```

Comment:
I am not sure of the `deleted_choice` variable name here. I get it that the option was deleted once but it has been added again.

Reply:
will fix this once the behavioural questions are resolved.

Reply:
resolving since this is outdated